osquery/filesystem/linux/proc.cpp

```diff
@@ -118,71 +118,211 @@ unsigned short procDecodePortFromHex(const std::string& encoded_port) {
   return decoded;
 }
 
-Status procProcesses(std::set<std::string>& processes) {
-  auto procProcessesCallback = [](const std::string& process_id,
-                                  std::set<std::string>& processes) -> bool {
-    processes.insert(process_id);
-    return true;
-  };
+static Status procGetSocketInodeToSocketInfoMapInet(
+    int family,
+    int protocol,
+    const std::string& path,
+    const std::string& content,
+    SocketInodeToSocketInfoMap& result) {
+  // The system's socket information is tokenized by line.
+  bool header = true;
+  for (const auto& line : osquery::split(content, "\n")) {
+    if (header) {
+      if (line.find("sl") != 0 && line.find("sk") != 0) {
+        return Status(1, std::string("Invalid file header for ") + path);
+      }
+      header = false;
+      continue;
+    }
 
-  return procProcesses<decltype(processes)>(procProcessesCallback, processes);
-}
+    // The socket information is tokenized by spaces, each a field.
+    auto fields = osquery::split(line, " ");
+    if (fields.size() < 10) {
+      VLOG(1) << "Invalid socket descriptor found: '" << line
+              << "'. Skipping this entry";
+      continue;
+    }
 
-Status procProcessSockets(std::vector<ProcessSocket>& socket_list,
-                          const std::string& process_id,
-                          int protocol,
-                          int family) {
-  socket_list.clear();
+    // Two of the fields are the local/remote address/port pairs.
+    auto locals = osquery::split(fields[1], ":");
+    auto remotes = osquery::split(fields[2], ":");
 
-  auto L_procProcessSocketsCallback =
-      [](const ProcessSocket& proc_socket,
-         std::vector<ProcessSocket>& socket_list) -> bool {
-    socket_list.push_back(proc_socket);
-    return true;
-  };
+    if (locals.size() != 2 || remotes.size() != 2) {
+      VLOG(1) << "Invalid socket descriptor found: '" << line
+              << "'. Skipping this entry";
+      continue;
+    }
+
+    SocketInfo socket_info = {};
+    socket_info.socket = fields[9];
+    socket_info.family = family;
+    socket_info.protocol = protocol;
+    socket_info.local_address = procDecodeAddressFromHex(locals[0], family);
+    socket_info.local_port = procDecodePortFromHex(locals[1]);
+    socket_info.remote_address = procDecodeAddressFromHex(remotes[0], family);
+    socket_info.remote_port = procDecodePortFromHex(remotes[1]);
+
+    if (protocol == IPPROTO_TCP) {
+      char* null_terminator_ptr = nullptr;
+      auto integer_socket_state =
+          std::strtoull(fields[3].data(), &null_terminator_ptr, 16);
```

Comment:
osquery's built in functions don't work for this?

Reply:
I didn't write this code and didn't want to change in this PR.